charms/openfga-k8s/src/charm.py

```diff
@@ -606,171 +590,157 @@ def _on_schema_upgrade_action(self, event):
-            try:
-                old_csr = self.state.get(STATE_KEY_CSR)
-                private_key = self.state.get(STATE_KEY_PRIVATE_KEY)
-            except RelationNotReadyError:
-                event.defer()
-                return
+            return
 
-            dnsname = "{}.{}-endpoints.{}.svc.cluster.local".format(
-                self.unit.name.replace("/", "-"),
-                self.app.name,
-                self.model.name,
-            )
-            if self.state.get(STATE_KEY_DNS_NAME):
-                dnsname = self.state.get(STATE_KEY_DNS_NAME)
+        if not self._state.is_ready():
+            logger.info("state not ready")
+            event.defer()
+            return
 
-            new_csr = generate_csr(
-                private_key=private_key.encode(), subject=dnsname
-            )
-            self.certificates.request_certificate_renewal(
-                old_certificate_signing_request=old_csr,
-                new_certificate_signing_request=new_csr,
-            )
-            try:
-                self.state.set(STATE_KEY_CSR, new_csr.decode())
-            except RelationNotReadyError:
-                event.defer()
-                return
+        old_csr = self._state.csr
+        private_key = self._state.private_key
+
+        dnsname = "{}.{}-endpoints.{}.svc.cluster.local".format(
+            self.unit.name.replace("/", "-"),
+            self.app.name,
+            self.model.name,
+        )
+        if self._state.dns_name:
+            dnsname = self._state.dns_name
+
+        new_csr = generate_csr(private_key=private_key.encode(), subject=dnsname)
+        self.certificates.request_certificate_renewal(
+            old_certificate_signing_request=old_csr,
+            new_certificate_signing_request=new_csr,
+        )
+
+        self._state.csr = new_csr.decode()
 
         self._update_workload()
 
     def _on_certificate_revoked(self, event: CertificateRevokedEvent) -> None:
         if not self.unit.is_leader():
-            old_csr = ""
-            private_key = ""
-            try:
-                old_csr = self.state.get(STATE_KEY_CSR)
-                private_key = self.state.get(STATE_KEY_PRIVATE_KEY)
-
-                dnsname = "{}.{}-endpoints.{}.svc.cluster.local".format(
-                    self.unit.name.replace("/", "-"),
-                    self.app.name,
-                    self.model.name,
-                )
-                if self.state.get(STATE_KEY_DNS_NAME):
-                    dnsname = self.state.get(STATE_KEY_DNS_NAME)
+            return
 
-                new_csr = generate_csr(
-                    private_key=private_key.encode(),
-                    subject=dnsname,
-                )
-                self.certificates.request_certificate_renewal(
-                    old_certificate_signing_request=old_csr,
-                    new_certificate_signing_request=new_csr,
-                )
+        if not self._state.is_ready():
```

Comment:
We could probably just make a decorator called "requires_state" that we add to all these functions.